common/remote/hooks/s3/s3file.cpp

```diff
@@ -16,700 +16,1320 @@
- */
-
-//#define TRACE_S3
-//#define TEST_S3_PAGING
-//#define FIXED_CREDENTIALS
-
-constexpr const char * s3FilePrefix = "s3://";
-
-#ifdef TEST_S3_PAGING
-constexpr offset_t awsReadRequestSize = 50;
-#else
-constexpr offset_t awsReadRequestSize = 0x400000;  // Default to requesting 4Mb each time
-#endif
 
-static std::atomic_bool initializedAws;
+using namespace Aws;
+
+// Constants
+constexpr const char* s3FilePrefix = "s3:";
+constexpr size_t s3FilePrefixLen = 3;  // Length of "s3:" for pointer arithmetic
+constexpr size32_t minMultipartSize = 5 * 1024 * 1024; // 5MB AWS minimum
+constexpr unsigned defaultMaxRetries = 3;
+
+// Global AWS initialization with reference counting
+static unsigned awsInitRefCount = 0;
 static CriticalSection awsCS;
-static Aws::SDKOptions options;
-MODULE_INIT(INIT_PRIORITY_HQLINTERNAL)
-{
-    return true;
-}
-MODULE_EXIT()
+static Aws::SDKOptions awsOptions;
+
+static void initAWS()
 {
-    if (initializedAws)
+    CriticalBlock block(awsCS);
+    if (awsInitRefCount == 0)
     {
-        Aws::ShutdownAPI(options);
-        initializedAws = false;
+        // First initialization
+        awsOptions.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Warn;
+        Aws::InitAPI(awsOptions);
+        PROGLOG("AWS SDK initialized for S3 file operations");
     }
+
+    awsInitRefCount++;
 }
 
-static void ensureAWSInitialized()
+static void shutdownAWS()
 {
-    if (initializedAws)
-        return;
     CriticalBlock block(awsCS);
-    if (initializedAws)
-        return;
-    Aws::InitAPI(options);
-    initializedAws = true;
-}
+    if (awsInitRefCount == 1)
+    {
+        // Last reference - perform shutdown
+        // IMPORTANT: All S3Clients must be destroyed before this point
+        Aws::ShutdownAPI(awsOptions);
+        PROGLOG("AWS SDK shutdown for S3 file operations");
+    }
+    else if (awsInitRefCount == 0)
+    {
+        // Should never happen - more shutdowns than inits
+        ERRLOG("AWS SDK shutdown called more times than init");
+    }
 
-//---------------------------------------------------------------------------------------------------------------------
+    awsInitRefCount--;
+}
 
-class S3File;
-class S3FileReadIO : implements CInterfaceOf<IFileIO>
+// S3 Client cache key based on plane and device
+struct S3ClientKey
 {
-public:
-    S3FileReadIO(S3File * _file, Aws::S3::Model::GetObjectOutcome & firstRead, FileIOStats & _stats);
+    StringAttr planeName;
+    unsigned device;
 
-    virtual size32_t read(offset_t pos, size32_t len, void * data) override;
-    virtual offset_t size() override;
-    virtual void close() override
+    bool operator==(const S3ClientKey& other) const
     {
-        //This could set a flag here to check for reading after close(), but I don't think any file read code
-        //ever calls close, and it would be harmless (and would complicate the rest of the code).
+        return (planeName == other.planeName) && (device == other.device);
```

Comment:
equality is not implemented for StringAttr - so need to call strsame().  Better to use std::string.